core/positionable_helpers.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Utility functions for positioning UI elements.
+ * @author kozbial@google.com (Monica Kozbial)
+ */
+'use strict';
+
+goog.provide('Blockly.uiPosition');
+
+goog.require('Blockly.Scrollbar');
+goog.require('Blockly.utils.Rect');
+goog.require('Blockly.utils.toolbox');
+
+goog.requireType('Blockly.MetricsManager');
+goog.requireType('Blockly.WorkspaceSvg');
+
+
+/**
+ * Enum for vertical positioning.
+ * @enum {number}
+ * @package
+ */
+Blockly.uiPosition.verticalPosition = {
+  TOP: 0,
+  BOTTOM: 1
+};
+
+/**
+ * Enum for horizontal positioning.
+ * @enum {number}
+ * @package
+ */
+Blockly.uiPosition.horizontalPosition = {
+  LEFT: 0,
+  RIGHT: 1
+};
+
+/**
+ * An object defining a horizontal and vertical positioning.
+ * @typedef {{
+ *   horizontal: !Blockly.uiPosition.horizontalPosition,
+ *   vertical: !Blockly.uiPosition.verticalPosition
+ * }}
+ * @package
+ */
+Blockly.uiPosition.Position;
+
+/**
+ * Enum for bump rules to use for dealing with collisions.
+ * @enum {number}
+ * @package
+ */
+Blockly.uiPosition.bumpDirection = {
+  UP: 0,
+  DOWN: 1
+};
+
+/**
+ * Returns a rectangle representing reasonable position for where to place a ui
+ * element of the specified size given the restraints and locations of the
+ * scrollbars. This method does not take into account any already placed ui
+ * elements.
+ * @param {!Blockly.uiPosition.Position} position The starting
+ *    horizontal and vertical position.
+ * @param {!Blockly.utils.Size} size the size of the ui element so get a start
+ *    position for.
+ * @param {number} horizontalPadding The horizontal padding to use. This value
+ *    is ignored for center horizontal positioning.
+ * @param {number} verticalPadding The vertical padding to use. This value
```

Comment:
Is the comment about center horizontal and vertical padding still valid? It looks like we removed the option to center positionables.

Reply:
Good catch. Yeah, it's not relevant anymore.